### PR TITLE
Replace Android ProGuard configuration with R8

### LINKED ISSUE #9941: Update builtin dmengine.pro
**Task (REQUIRED):**
Builtin dmengine.pro Proguard definition produces build error.
Investigate the root cause and fix it

Report on forum https://forum.defold.com/t/defold-1-9-6-beta/79030/43

**Expected outcome (REQUIRED):**
No errors, everything builds fine.

### LINKED ISSUE #12976: 🟡 Java code obfuscation/minification and optimization using R8
https://developer.android.com/topic/performance/app-optimization/enable-app-optimization

### PR DESCRIPTION
**Pull request by AGulev**
Android projects can now enable Java shrinking, optimization, and obfuscation with R8 by selecting a `.keep` rules file. Projects without one continue using D8, while the new built-in `dmengine.keep` provides safe defaults without the errors caused by the obsolete ProGuard configuration.

Fix https://github.com/defold/defold/issues/9941
Fix https://github.com/defold/defold/issues/12976

### Technical changes

- Replace `android.proguard` with the Android-only `android.r8_keep_rules` setting.
- Upload selected rules as `_app/app.keep` and trigger an Extender build when R8 is enabled.
- Replace the ProGuard Extender configuration with R8, including AAPT2-generated rules and mapping output.
- Retain D8 as the default when no R8 rules are configured.
- Replace `dmengine.pro` with concise built-in rules that preserve Defold’s JNI and reflection entry points.
- Copy `mapping.txt` into the APK symbols directory for symbol-enabled builds.
- Update Editor labels, help text, localization, and project fixtures.
- Existing `android.proguard` settings no longer enable optimization; projects must migrate to a `.keep` file referenced by `android.r8_keep_rules`.
- Add tests covering built-in, custom, empty, missing, legacy, and platform-specific configurations, plus mapping-file copying.
- Validate with the focused `ExtenderUtilTest` and `AndroidBundlerR8MappingTest` suites.

### CODE REVIEW

File: share/extender/build_input.yml [hunk 1]
```diff
@@ -397,7 +397,8 @@ platforms:
 
   android:
     env:
-        PROGUARD:                 "{{env.ANDROID_PROGUARD}}"
+        R8:                       "{{env.ANDROID_R8}}"
+        R8_VERSION:               "{{env.ANDROID_R8_VERSION}}"
```
Comment (AGulev) [inline share/extender/build_input.yml:403]: build fail with 
```
Cannot create property=r8Cmd for JavaBean=com.defold.extender.PlatformConfig@2c0a794c
```
what the fix here? merge extender first, then run it on stage and only then merge the engine?

File: engine/engine/content/builtins/manifests/android/dmengine.keep [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# R8 keep rules for the Java classes shipped with the Defold engine.
```
[the remark this answers is not in the document]
Comment (AGulev) [reply engine/engine/content/builtins/manifests/android/dmengine.keep:1]: we don't care about WAF anymore

**Comment by AGulev** [comment]
@codex review

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/AndroidBundler.java [hunk 3]
```diff
@@ -738,6 +734,14 @@ private void signFile(Project project, File signFile, ICanceled canceled) throws
     /**
     * Copy debug symbols
     */
+    static void copyR8Mapping(File extenderExeDir, Platform architecture, File symbolsDir) throws IOException {
+        File r8Mapping = new File(new File(extenderExeDir, architecture.getExtenderPair()), "mapping.txt");
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/AndroidBundler.java]:
```java
File r8Mapping = Path.of(extenderExeDir.getPath(), architecture.getExtenderPair(), "mapping.txt").toFile();
```

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties [hunk 4]
```diff
@@ -839,12 +839,12 @@ debuggable.type = bool
...
 debuggable.default = 0
 
-proguard.type = resource
-proguard.help = custom ProGuard file to help strip redundant Java classes from the final .apk
-proguard.default =
-proguard.preserve-extension = 1
-proguard.filter = pro
-proguard.label = ProGuard config
+r8_keep_rules.type = resource
+r8_keep_rules.help = Complete R8 keep-rule file used when shrinking, optimizing, and obfuscating Java code in the Android build
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties]: What do we mean by "Complete"?

**Review by britzl — COMMENTED** [review]
I had a question and a comment only

**Comment by AGulev** [comment]
I also prepared fixes for the following extensions, but I'll update them only when both PRs will be merged (and tested al lof them on a local extender):
AdInfo
AdMob
Camera
DefVideoAds
Facebook
Firebase
Firebase Analytics
Firebase Remote Config
Firebase Crashlytics
GPGS
IAP
Push
Review
VideoPlayer
WebView
Sharing
LevelPlay
Zendesk
Permissions

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/AndroidBundler.java (hunk 3 above)
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/AndroidBundler.java]: fixed

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties (hunk 4 above)
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties]: fixed

File: editor/resources/localization/en.editor_localization [hunk 5]
```diff
@@ -1964,8 +1964,8 @@ form.label.project.android.display_cutout = Display Cutout
...
 form.label.project.android.debuggable = Debuggable
 form.help.project.android.debuggable = Sets `android:debuggable` in the manifest, allowing tools such as Android Studio profilers and debuggers to inspect the application and its files. Leave disabled for release builds unless explicitly needed.
-form.label.project.android.proguard = ProGuard Config
-form.help.project.android.proguard = Custom `.pro` rules used when shrinking and obfuscating Java code in the Android build. Add keep rules for code accessed through reflection or JNI so it is not removed or renamed. Leave empty when no project-specific rules are needed.
+form.label.project.android.r8_keep_rules = R8 Keep Rules
+form.help.project.android.r8_keep_rules = A `.keep` file that enables R8 shrinking, optimization, and obfuscation for the Android build. The selected file is the complete project rule set; copy the built-in `dmengine.keep` before customizing it. Leave empty to use D8 without shrinking.
```
Comment (ekharkunov) [inline editor/resources/localization/en.editor_localization]: I think it's better to phrase like `Leave empty to get final build without shrinking.` because it can be some mismatch between D8 and R8 for users who is not familiar with Android internals

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ExtenderUtil.java [hunk 6]
```diff
@@ -562,6 +562,10 @@ private static boolean hasPropertyResource(Project project, BobProjectProperties
         return false;
     }
 
+    private static boolean isAndroidPlatform(Platform platform) {
+        return platform != null && "android".equals(platform.getOs());
```
Comment (ekharkunov) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ExtenderUtil.java]: I think `platform != null` check is redundant here because if `getPlatform()` is null it throws Exception and build stops earlier.

**Review by ekharkunov — DISMISSED** [review]
Overall good. See comments

File: editor/resources/localization/en.editor_localization (hunk 5 above)
Comment (AGulev) [reply editor/resources/localization/en.editor_localization]: thank you! Fixed!

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ExtenderUtil.java (hunk 6 above)
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ExtenderUtil.java]: Fixed!

**Review by ekharkunov — APPROVED** [review]

**Comment by AGulev** [comment]
@ekharkunov  As I understand we shouldn't merge it until we have a stage server with needed changes?

**Comment by ekharkunov** [comment]
> @ekharkunov As I understand we shouldn't merge it until we have a stage server with needed changes?

Correct

**Comment by AGulev** [comment]
@ekharkunov do we have estimation when stage server will be ready so I can merge it?

**Comment by ekharkunov** [comment]
> @ekharkunov do we have estimation when stage server will be ready so I can merge it?

End of the week.
